### PR TITLE
[CSE 276F Submission] PushT-v1

### PR DESCRIPTION
**Pull request by Xander-Hinrichsen**
PushT from diffusion policy digital twin

Created PandaStick agent
Fully batched calculation for T block intersection area with goal T area, zero explicit loops

Support for ur5e robot   is a todo for true twinness
Wristcam attached to PandaStick is a todo for visual based RL/IL

ppo args (consistent learning):
python ppo.py --env_id="PushT-v1" --exp-name="final_run" --num_envs=1024 --update_epochs=8 --num_minibatches=32   --total_timesteps=55_000_000 --eval_freq=8 --num-steps=100 --num_eval_steps=100 --gamma=0.99

### CODE REVIEW

**Comment by Xander-Hinrichsen** [comment]
@StoneT2000 , could I request a review on this PR please?
Thank you!

File: docs/source/tasks/digital_twins/index.md [hunk 1]
```diff
@@ -1 +1,32 @@
-# Digital Twins (WIP)
\ No newline at end of file
+# Digital Twins (WIP)
+[asset-badge]: https://img.shields.io/badge/download%20asset-yes-blue.svg
+[reward-badge]: https://img.shields.io/badge/dense%20reward-yes-green.svg
+
+## PushT-v1
```
Comment (StoneT2000) [inline docs/source/tasks/digital_twins/index.md]: actually for now lets just put this under the table-top tasks section (it is named for two finger grippers, that will be fixed). We want to leave this section for tested digital twins with real world comparisons

File: mani_skill/envs/tasks/digital_twins/push_T.py [hunk 2]
```diff
@@ -0,0 +1,509 @@
...
+from mani_skill.envs.sapien_env import BaseEnv
+from mani_skill.sensors.camera import CameraConfig
+from mani_skill.utils import common, sapien_utils
+from mani_skill.utils.building import actors
+from mani_skill.utils.registration import register_env
+from mani_skill.utils.scene_builder.table import TableSceneBuilder
+from mani_skill.utils.structs import Pose
+from mani_skill.utils.structs.types import Array, GPUMemoryConfig, SimConfig
+
+##additional imports - for making T block
```
Comment (StoneT2000) [inline mani_skill/envs/tasks/digital_twins/push_T.py]: remove this comment

File: mani_skill/envs/tasks/digital_twins/push_T.py [hunk 3]
```diff
@@ -0,0 +1,509 @@
...
+import sapien 
+
+##extending TableSceneBuilder and only making 2 changes:
+##1.Making table smooth and white, 2. adding support for keyframes of new robots - panda stick
```
Comment (StoneT2000) [inline mani_skill/envs/tasks/digital_twins/push_T.py]: this comment is fine but use the standard "# my comment" format instead of "##comment here"

File: mani_skill/envs/tasks/digital_twins/push_T.py [hunk 4]
```diff
@@ -0,0 +1,509 @@
...
+
+@register_env("PushT-v1", max_episode_steps=100)
+class PushTEnv(BaseEnv):
+    """
+    Task Description
+    ----------------
+    Easier Digital Twin of real life push-T task from Diffusion Policy: https://diffusion-policy.cs.columbia.edu/
+    "In this task, the robot needs to
+    ① precisely push the T- shaped block into the target region, and
+    ② move the end-effector to the end-zone which terminates the episode. [② Not required for PushT-easy-v1]"
```
Comment (StoneT2000) [inline mani_skill/envs/tasks/digital_twins/push_T.py]: try to avoid using non standard characters, stick to 1. 2.

File: mani_skill/envs/tasks/digital_twins/push_T.py [hunk 5]
```diff
@@ -0,0 +1,509 @@
...
+    - Table View Camera parameters          sapien_utils.look_at(eye=[0.3, 0, 0.6], target=[-0.1, 0, 0.1])
+    
+    TODO's
+    ------
+    - Add hand mounted camera for panda_stick robot, for visual rl
+    - Add support for ur5e robot with hand mounted camera and real life end effector (3D cad link in their github README)
+    - Tune Unspecified Real-World Parameters
+    - Add robot qpos to randomizations
+
+    Visualization: TODO, example https://maniskill.readthedocs.io/en/latest/tasks/index.html#pushcube-v1
```
Comment (StoneT2000) [inline mani_skill/envs/tasks/digital_twins/push_T.py]: for TODO comments we typically write them as

"# TODO (short user id): comment"

E.g. for me I use "# TODO (stao): comment". Please change your TODO comments as so others know who wrote this task

File: mani_skill/envs/tasks/digital_twins/push_T.py [hunk 6]
```diff
@@ -0,0 +1,509 @@
...
+                height=128,
+                fov=np.pi / 2,
+                near=0.01,
+                far=100,
+            )
+        ]
+
+    @property
+    def _default_human_render_camera_configs(self):
+        # registers a more high-definition (512x512) camera used just for rendering when render_mode="rgb_array" or calling env.render_rgb_array()
```
Comment (StoneT2000) [inline mani_skill/envs/tasks/digital_twins/push_T.py]: you can remove this comment, its from the tutorial

File: mani_skill/envs/tasks/digital_twins/push_T.py [hunk 7]
```diff
@@ -0,0 +1,509 @@
...
+        self.table_scene = WhiteTableSceneBuilder(
+            env=self, robot_init_qpos_noise=self.robot_init_qpos_noise
+        )
+        self.table_scene.build()
+        
+        #returns 3d cad of create_tee - center of mass at (0,0,0)
+        ## cad Tee is upside down (⊥) (both 3D tee and target)
+        TARGET_RED = np.array([194, 19, 22, 255]) / 255 ##same as mani_skill.utils.building.actors.common - goal target
+        def create_tee(name="tee", target=False, base_color=TARGET_RED):
+                ##dimensions of boxes that make tee
```
Comment (StoneT2000) [inline mani_skill/envs/tasks/digital_twins/push_T.py]: fix this indenting, it is indented twice in here.

File: mani_skill/envs/tasks/digital_twins/push_T.py [hunk 8]
```diff
@@ -0,0 +1,509 @@
...
+
+        ############## Rest of function is Custom 2D "Pseudo-Rendering" function for fully batched GPU T Intersection Area Calculation ##############
+        res = 64
+        uv_half_width = 0.15
+        self.uv_half_width = uv_half_width
+        self.res = res
+        oned_grid = (torch.arange(res, dtype=torch.float32).view(1,res).repeat(res,1) - (res/2))
+        self.uv_grid = (torch.cat([oned_grid.unsqueeze(0), (-1*oned_grid.T).unsqueeze(0)], dim=0) + 0.5) / ((res/2)/uv_half_width)
+        self.uv_grid = self.uv_grid.to(self.device)
+        self.homo_uv = torch.cat([self.uv_grid, torch.ones_like(self.uv_grid[0]).unsqueeze(0)], dim=0)
```
Comment (StoneT2000) [inline mani_skill/envs/tasks/digital_twins/push_T.py]: :dizzy_face: this is impressive work here, this is the only GPU parallelized push T task in the world then. I checked original diffusion policy repo + Le Robot repo and both are using external cpu based package.

File: mani_skill/envs/tasks/digital_twins/push_T.py [hunk 9]
```diff
@@ -0,0 +1,509 @@
...
+        box2 = box2.long()
+
+        self.tee_render = torch.zeros(res,res)
+        ##image map has flipped x and y, set values in transpose to undo
+        self.tee_render.T[box1[0,0]:box1[1,0], box1[2,1]:box1[0,1]] = 1
+        self.tee_render.T[box2[0,0]:box2[1,0], box2[2,1]:box2[0,1]] = 1
+        #image map y is flipped of xy plane, flip to unflip
+        self.tee_render = self.tee_render.flip(0).to(self.device)
+
+        def quat_to_z_euler(quats):
```
Comment (StoneT2000) [inline mani_skill/envs/tasks/digital_twins/push_T.py]: these functions (i did not verify) may be useful to be part of the standard geometry submodule in maniskill in the future. Can leave it here for now though.

File: mani_skill/envs/tasks/digital_twins/push_T.py [hunk 10]
```diff
@@ -0,0 +1,509 @@
...
+        zrot = self.quat_to_zrot(goal_fake_quat).squeeze(0) # 3x3 rot matrix for goal to world transform
+        goal_trans = torch.eye(3)
+        goal_trans[:2,:2] = zrot[:2,:2]
+        goal_trans[0:2, 2] = self.goal_offset
+        self.world_to_goal_trans = torch.linalg.inv(goal_trans).to(self.device) ##this is just a 3x3 matrix (2d homogenious transform)
+        
+        ##this is the 'pseudo render' algo for calculating the intersection
+        ##made custom 'psuedo renderer' to compute intersection area in parallel on cuda, zero explicit loops
+        ##just views blocks in 2d in the goal tee frame to see overlap, fairly cheap operation
+        def pseudo_render_intersection():
```
Comment (StoneT2000) [inline mani_skill/envs/tasks/digital_twins/push_T.py]: why is this function defined locally? Can this not be defined in a more object oriented manner without what might look like side effects (e.g. accessing and modifying self with a function that is not given the variable self).

File: mani_skill/envs/tasks/digital_twins/push_T.py [hunk 11]
```diff
@@ -0,0 +1,509 @@
...
+
+    def _initialize_episode(self, env_idx: torch.Tensor, options: dict):
+        # use the torch.device context manager to automatically create tensors on CPU or CUDA depending on self.device, the device the environment runs on
+        with torch.device(self.device):
+            # the initialization functions where you as a user place all the objects and initialize their properties
+            # are designed to support partial resets, where you generate initial state for a subset of the environments.
+            # this is done by using the env_idx variable, which also tells you the batch size
+            b = len(env_idx)
+            # when using scene builders, you must always call .initialize on them so they can set the correct poses of objects in the prebuilt scene
+            # note that the table scene is built such that z=0 is the surface of the table.
```
Comment (StoneT2000) [inline mani_skill/envs/tasks/digital_twins/push_T.py]: please remove most of the tutorial comments

File: mani_skill/envs/tasks/digital_twins/push_T.py [hunk 12]
```diff
@@ -0,0 +1,509 @@
...
+                Pose.create_from_pq(
+                    p=xyz,
+                    q=euler2quat(0, np.pi / 2, 0),
+                )
+            )
+
+    def evaluate(self):
+        ##success is where the overlap is over intersection thresh and ee dist to start pos is less than it's own thresh
+        inter_area = self.pseudo_render_intersection()
+        tee_place_success = (inter_area) >= self.intersection_thresh
```
Comment (StoneT2000) [inline mani_skill/envs/tasks/digital_twins/push_T.py]: how come the intersection area is not used for dense reward computation? Did you find this to be less performant compared to the implementation you have now?

File: mani_skill/envs/tasks/digital_twins/push_T.py [hunk 13]
```diff
@@ -0,0 +1,509 @@
...
+                obj_pose=self.tee.pose.raw_pose,
+                #ee_restart_goal_pos=self.ee_goal_pos.pose.p ##for full implementation
+            )
+        return obs
+
+    def compute_dense_reward(self, obs: Any, action: Array, info: Dict):
+        ##reward for overlap of the tees
+
+        #legacy reward
+        #reward = self.pseudo_render_reward()
```
Comment (StoneT2000) [inline mani_skill/envs/tasks/digital_twins/push_T.py]: leave a comment maybe about why the original diffusion policy paper reward is not used?

File: mani_skill/envs/tasks/digital_twins/push_T.py (hunk 12 above)
Comment (Xander-Hinrichsen) [reply mani_skill/envs/tasks/digital_twins/push_T.py]: Yes, it's far less performant. Every time I used the legacy reward function from the 2D environment, the agent found local max of pushing the horizontal part of the T onto the vertical part of the goal T to maximize intersection asap, and with a lack of exploration it just got stuck there (for cases where initial rotation of T was around 90/270 degrees compared to goal). Working on other finals, but will make the other fixes soon

This was in earlier testing of my code though. I will retry the legacy reward and compare for my report

also will remove comments talking about ee end position success - since it's no longer part of the environment

File: mani_skill/envs/tasks/digital_twins/push_T.py
Comment (StoneT2000) [inline mani_skill/envs/tasks/digital_twins/push_T.py:1]: change the T to lowercase t for python file naming consistency. Environment ID used atm is fine though.

**Review by StoneT2000 — CHANGES_REQUESTED** [review]
nice work, would love to get this merged in, just see the comments. ppo works fast as well.

File: mani_skill/envs/tasks/digital_twins/push_T.py
Comment (StoneT2000) [reply mani_skill/envs/tasks/digital_twins/push_T.py:1]: also sorry also move this to the tabletop folder in envs/tasks

**Comment by Xander-Hinrichsen** [comment]
@StoneT2000 , I've made the changes. Let me know if there are any more  issues

**Review by StoneT2000 — APPROVED** [review]
only some small issues with comments left but i will fix that myself. This looks good to merge
